### ISSUE TITLE
Suboptimal assembly for struct equality

### ISSUE THREAD

**Issue by rafihayne** [ziglang/zig#8391]
I stumbled across [this issue](https://github.com/rust-lang/rust/issues/83623) in the rust compiler and started playing with it in zig. With some help of @lithdew, it appears that similarly to rust, zig does not vectorize the function call. We tried with `std.meta.eql`, but the current zig implementation seems like a much more simple reproduction case

[Zig Godbolt](https://godbolt.org/z/Kq5W6nWbP)
[C++ Godbolt](https://godbolt.org/z/MYff5bPrP)

Unfortunately I don't know enough about compilers to debug further. If folks have an idea for how to dig into this, happy to learn and give it a shot with some advice.

**Comment by lithdew** [comment]
The issue appears to come down to how _passing by reference_ and _passing by pointer_ differs in C++. Zig automatically assumes passing by pointer for function arguments, and therefore can't vectorize as well in comparison to passing by reference like in C++.

Here's the difference in assembly between passing by reference and passing by pointer in C++: https://godbolt.org/z/WYc1W489r

**Comment by LemonBoy** [comment]
> The issue appears to come down to how passing by reference and passing by pointer differs in C++.

Not really. The problem is that the emitted IR is too convoluted and is failing to trigger the `mergeicmps` pass, Clang's IR manages to get all the comparisons to be folded into a single memcmp that's then expanded into vectorized ops.
